crates/noirc_evaluator/src/ssa/acir_gen/internal_var.rs

```diff
@@ -0,0 +1,233 @@
+use crate::{
+    ssa::acir_gen::{
+        const_from_expression, constraints::ACIRState, expression_to_witness,
+        optional_expression_to_witness,
+    },
+    ssa::node::NodeId,
+};
+use acvm::{
+    acir::native_types::{Expression, Witness},
+    FieldElement,
+};
+
+#[derive(Clone, Debug, Eq)]
+pub struct InternalVar {
+    // A multi-variate degree-2 polynomial
+    expression: Expression,
+    // A witness associated to `expression`.
+    // semantically `cached_witness` and `expression`
+    // should be the equal.
+    //
+    // Example: `z = x + y`
+    // `z` can be seen as the same to `x + y`
+    // ie if `z = 10` , then `x+y` must be equal to
+    // 10.
+    // There are places where we can use `cached_witness`
+    // in place of `expression` for performance reasons
+    // due to the fact that `cached_witness` is a single variable
+    // whereas `expression` is a multi-variate polynomial which can
+    // contain many degree-2 terms.
+    //
+    // Note that we also protect against the case that `expression`
+    // changes and `cached_witness` becomes "dirty" or outdated.
+    // This is because one is not allowed to modify the `expression`
+    // once set. Additionally, there are no methods that allow one
+    // to modify the `cached_witness`
+    cached_witness: Option<Witness>,
+    id: Option<NodeId>,
+}
+
+impl InternalVar {
+    pub(crate) fn expression(&self) -> &Expression {
+        &self.expression
+    }
+    pub(crate) fn set_id(&mut self, id: NodeId) {
+        match self.id {
+            Some(_existing_id) => {
+                // TODO: this assert will panic the compiler.
+                // TODO why does the NodeId change (Guillaume)?
+                // assert_eq!(existing_id, id, "ICE: changing node id to a different value")
+                self.id = Some(id)
+            }
+            None => self.id = Some(id),
```

Comment:
We do the same thing in all branches, so we can remove the match entirely:
```rs
// insert comment explaining this assert will panic, etc
// assert_eq!(self.id, None);
self.id = Some(id);
```

Reply:
Yeah, I tried to make it follow the previous way but add a compiler panic. I will remove it and create an issue regarding it